### PR TITLE
DP1.3: Adding deviation code for Cisco ecn_same_min_max_threshold_unsupproted

### PR DESCRIPTION
**Pull request by suredhar**
DP-1.3 test is failing with "'DPA_QOSEA' detected the 'warning' condition 'Invalid WRED configuration - Min value is greater than Max'" error due to changes in the test.

it doesn't support Cisco. minimum and maximum Threshold values can't be the same, as well as it should be a multiple of 6,144 bytes.

for that added the deviation code - ecn_same_min_max_threshold_unsupproted

### CODE REVIEW

File: internal/deviations/deviations.go [hunk 1]
```diff
@@ -625,3 +625,9 @@ func ISISLspMetadataLeafsUnsupported(dut *ondatra.DUTDevice) bool {
 func QOSQueueRequiresID(dut *ondatra.DUTDevice) bool {
 	return lookupDUTDeviations(dut).GetQosQueueRequiresId()
 }
+
+// EcnMaxMimumThreshold returns the maximum threshold value while comparing for pass or fail conditions.
+// Maximum and mimium threshold value not be same. Anything less Maximum threshold should be 6144 greate then mimnimum threshold value.
+func EcnSameMaxMimumThresholdUnsupported(dut *ondatra.DUTDevice) bool {
```
Comment (guoshiuan) [inline internal/deviations/deviations.go:631]: Should we name this function "EcnSameMinMaxThresholdUnsupproted"? and please also update the comment.

File: feature/qos/tests/qos_ecn_config_test/metadata.textproto [hunk 2]
```diff
@@ -14,3 +14,11 @@ platform_exceptions: {
     drop_weight_leaves_unsupported: true
   }
 }
+platform_exceptions: {
+  platform: {
+    vendor: CISCO
+  }
+  deviations: {
+    ecn_same_min_max_threshold_unsupproted: true
```
Comment (nraghavendran) [inline feature/qos/tests/qos_ecn_config_test/metadata.textproto:22]: pls correct spelling

Comment (suredhar) [reply feature/qos/tests/qos_ecn_config_test/metadata.textproto:22]: okay will correct it

File: internal/deviations/deviations.go (hunk 1 above)
Comment (suredhar) [reply internal/deviations/deviations.go:631]: okay

File: feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go [hunk 3]
```diff
@@ -36,6 +36,11 @@ type Testcase struct {
 	fn   func(t *testing.T)
 }
 
+const (
```
[the remark this answers is not in the document]
Comment (suredhar) [reply feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go:39]: comments added

File: internal/deviations/deviations.go (hunk 1 above)
Comment (guoshiuan) [reply internal/deviations/deviations.go:632]: The deviation function name should be the Camel case of the protobuf field name. For example, the func name of field "ecn_same_min_max_threshold_unsupported" is "EcnSameMinMaxThresholdUnsupported".

File: internal/deviations/deviations.go [hunk 4]
```diff
@@ -625,3 +625,10 @@ func ISISLspMetadataLeafsUnsupported(dut *ondatra.DUTDevice) bool {
 func QOSQueueRequiresID(dut *ondatra.DUTDevice) bool {
 	return lookupDUTDeviations(dut).GetQosQueueRequiresId()
 }
+
+// GetEcnSameMinMaxThresholdUnsupported returns true for devices that don't support Minimum and Maximum threshold values are same
```
Comment (guoshiuan) [inline internal/deviations/deviations.go:629]: The comment of Go function should start with the function name, i.e., "EcnSameMinMaxThresholdUnsupported returns..." in this case.

File: internal/deviations/deviations.go [hunk 5]
```diff
@@ -625,3 +625,10 @@ func ISISLspMetadataLeafsUnsupported(dut *ondatra.DUTDevice) bool {
...
+// EcnMaxMimumThreshold returns the maximum threshold value while comparing for pass or fail conditions.
```
Comment (guoshiuan) [inline internal/deviations/deviations.go:630]: I didn't see this function "EcnMaxMimumThreshold" defined.

File: internal/deviations/deviations.go [hunk 6]
```diff
@@ -625,3 +625,10 @@ func ISISLspMetadataLeafsUnsupported(dut *ondatra.DUTDevice) bool {
...
+// CISCO: Maximum and mimium threshold value not be same. Anything less Maximum threshold should be 6144 greate then mimnimum threshold value.
```
Comment (guoshiuan) [inline internal/deviations/deviations.go:631]: Lots of typos here, please revise.

s/mimimu/minimum
s/less Maximum/less than maximum
s/greate then mimnimum/greater than minimum

File: feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go [hunk 7]
```diff
@@ -36,6 +36,13 @@ type Testcase struct {
 	fn   func(t *testing.T)
 }
 
+// For Cisco devices, minimum and maximum Threshold values can't be the same,
```
Comment (guoshiuan) [inline feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go:39]: s/Threshold/threshold

Comment (suredhar) [reply feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go:39]: modified

File: internal/deviations/deviations.go (hunk 6 above)
Comment (suredhar) [reply internal/deviations/deviations.go:631]: revised

File: internal/deviations/deviations.go (hunk 5 above)
Comment (suredhar) [reply internal/deviations/deviations.go:630]: corrected

File: internal/deviations/deviations.go [hunk 8]
```diff
@@ -625,3 +625,9 @@ func ISISLspMetadataLeafsUnsupported(dut *ondatra.DUTDevice) bool {
 func QOSQueueRequiresID(dut *ondatra.DUTDevice) bool {
 	return lookupDUTDeviations(dut).GetQosQueueRequiresId()
 }
+
+// EcnSameMinMaxThresholdUnsupported returns true for devices that don't support Minimum and Maximum threshold values are same
```
Comment (guoshiuan) [inline internal/deviations/deviations.go:629]: ... that don't support the same value for Minimum and Maximum threshold.

File: internal/deviations/deviations.go [hunk 9]
```diff
@@ -625,3 +625,9 @@ func ISISLspMetadataLeafsUnsupported(dut *ondatra.DUTDevice) bool {
...
+// CISCO: minimum and maximum threshold value not be same. Anything less than maximum threshold should be 6144 greater than minimum threshold value.
```
Comment (guoshiuan) [inline internal/deviations/deviations.go:630]: ... values cannot be the same; anything ...

File: internal/deviations/deviations.go (hunk 1 above)
Comment (suredhar) [reply internal/deviations/deviations.go]: modified

File: internal/deviations/deviations.go (hunk 4 above)
Comment (suredhar) [reply internal/deviations/deviations.go]: corrected

File: feature/qos/tests/qos_ecn_config_test/metadata.textproto [hunk 10]
```diff
@@ -14,3 +14,11 @@ platform_exceptions: {
...
   }
 }
+platform_exceptions: {
+  platform: {
+    vendor: CISCO
+  }
+  deviations: {
+    ecn_same_min_max_threshold_unsupported: true
+  }
+}
```
Comment (Ankur19) [inline feature/qos/tests/qos_ecn_config_test/metadata.textproto:24]: Add new line here

File: feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto [hunk 11]
```diff
@@ -1,20 +1,19 @@
...
 # proto-message: Metadata
 
-uuid: "b960ced4-2f59-49bf-b078-e53f169fac06"
-plan_id: "RT-1.19"
-description: "BGP 2-Byte and 4-Byte ASN support"
-testbed: TESTBED_DUT_ATE_2LINKS
-platform_exceptions: {
-  platform: {
-    vendor: ARISTA
+uuid:  "b960ced4-2f59-49bf-b078-e53f169fac06"
```
Comment (Ankur19) [inline feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto:4]: Revert all these to old spacing?

File: feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto [hunk 12]
```diff
@@ -1,20 +1,19 @@
...
-    network_instance_table_deletion_required: true
+  deviations:  {
+    route_policy_under_afi_unsupported:  true
+    omit_l2_mtu:  true
+    network_instance_table_deletion_required:  true
+    interface_enabled:  true
+    default_network_instance:  "default"
   }
 }
-tags: TAGS_AGGREGATION
```
Comment (mohanasm) [inline feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto:20]: Any reason for removing the tags in all the metadata.textproto file? Can you revert this change?

**Review by mohanasm — CHANGES_REQUESTED** [review]

File: feature/qos/tests/qos_ecn_config_test/metadata.textproto (hunk 10 above)
Comment (suredhar) [reply feature/qos/tests/qos_ecn_config_test/metadata.textproto:24]: added

File: internal/deviations/deviations.go (hunk 9 above)
Comment (suredhar) [reply internal/deviations/deviations.go]: modified

File: internal/deviations/deviations.go (hunk 8 above)
Comment (suredhar) [reply internal/deviations/deviations.go]: modified

File: feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go [hunk 13]
```diff
@@ -471,12 +485,21 @@ func testCiscoECNConfig(t *testing.T) {
 	if got, want := gnmi.GetConfig(t, dut, wredUniform.EnableEcn().Config()), ecnConfig.ecnEnabled; got != want {
 		t.Errorf("wredUniform.EnableEcn().State(): got %v, want %v", got, want)
 	}
+	if deviations.EcnSameMinMaxThresholdUnsupported(dut) {
```
Comment (Ankur19) [inline feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go:488]: Can we create two variables here for wantMinThreshold and wantMaxThreshold and apply the deviation logic to those? We need not copy the whole if/else statements

**Review by Ankur19 — DISMISSED** [review]
LGTM! 

nit: remove the duplicate if/else error statements and just create variables with the wanted value.

File: feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto (hunk 11 above)
Comment (suredhar) [reply feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto]: done

File: feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto (hunk 12 above)
Comment (suredhar) [reply feature/experimental/bgp/ate_tests/bgp_2byte_4byte_asn/metadata.textproto]: no,

revert back it

File: feature/qos/tests/qos_ecn_config_test/metadata.textproto (hunk 2 above)
Comment (suredhar) [reply feature/qos/tests/qos_ecn_config_test/metadata.textproto]: corrected

File: feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go (hunk 13 above)
Comment (suredhar) [reply feature/qos/tests/qos_ecn_config_test/qos_ecn_config_test.go]: modified

**Comment by suredhar** [comment]
closing this PR because of  "DP-1.3: Adding test support for Nokia #2022"
